### DISCUSSION TITLE
No clear guidance on how to create a mock route context with body

### DISCUSSION THREAD

**Discussion by Skwai** [oakserver/oak#633] [Q&A]
I'm trying to write tests for a route I've created. However I cannot for the life of me figure out how to set the `request.body` on `createMockContext`. The `request.body` is readonly so clearly it's not intended to be mutated like so:

```
const ctx = createMockContext()
ctx.request.body = "foo"
```

I've tried so search the codebase but can't see any examples or interfaces for creating a mocked JSON request context.

**Comment by jefgodesky** [comment]
`createMockContext` expects you to pass in the stuff you want in the context as parameters, so you wouldn't create the context and then assign the body; rather, you pass the body in to `createMockContext`, like this:

```typescript
const ctx = createMockContext({
   body: 'foo'
})
```

Except no, that isn't quite right either, because `body` isn't a string, it's a `ReadableStream<any>`. So to help with that, I wrote a utility method called `stringToReadableStream`. But I also favor test-driven development (TDD), which meant writing a test first, so:

```typescript
import { describe, it } from 'jsr:@std/testing/bdd'
import { expect } from 'jsr:@std/expect'
import stringToReadableStream from './string-to-readable-stream.ts'

describe('stringToReadableStream', () => {
  const greeting = 'Hello, world!'
  const readableStreamToString = async (stream: ReadableStream<any>): Promise<string> => {
    const reader = stream.getReader()
    const decoder = new TextDecoder()
    let result = ''

    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      result += decoder.decode(value, { stream: true })
    }

    result += decoder.decode()
    return result
  }

  it('returns a readable stream', async () => {
    const stream = stringToReadableStream(greeting)
    const back = await readableStreamToString(stream)
    expect(typeof stream).not.toBe('string')
    expect(back).toEqual(greeting)
  })
})
```

_Then_...

```typescript
const stringToReadableStream = (orig: string): ReadableStream<any> => {
  return new ReadableStream({
    start (controller) {
      controller.enqueue(new TextEncoder().encode(orig))
      controller.close()
    }
  })
}

export default stringToReadableStream
```

At which point you can bring it all together:

```typescript
import stringToReadableStream from './utils/string-to-readable-stream.ts'

const ctx = createMockContext({
   body: stringToReadableStream('foo')
})
```

And _that_ works!
